### ISSUE TITLE
Warning for assignment of mutables

### ISSUE THREAD

**Issue by boolean-light** [astral-sh/ruff#10507]
### Discussed in https://github.com/astral-sh/ruff/discussions/10408

<div type='discussions-op-text'>

<sup>Originally posted by **boolean-light** March 14, 2024</sup>
I've run into lots of bugs related to assignments of mutables (e.g. lists, dicts, sets), like this:
```python
array = [1, 2, 3]
temp_array = array
temp_array[0] = 42  # This will change content of original array also!
print(temp_array)  # [42, 2, 3]
print(array)  # [42, 2, 3]
```
But as far as I know, there are no rules which gives warning about this.
Could we add warning or possibly fixes (e.g. suggesting `copy` or `deepcopy`)?

Probably because it requires type inference, it will be difficult to make it work for something like this:
```python
array = [1, 2, 3]  # This is not intended to be modified

def change_first(arr):
    arr[0] = 42
    return arr

add_first(array)  # This will change the array
print(array)  # [42, 2, 3]
```
But I think we could do this, like for first example, where the type of variable is straightforward.</div>

Thought this could be opened as an issue, related to #10505.

**Comment by MichaReiser** [comment]
I understand that the rule you're asking for is to warn if multiple variables point to the same data and that data gets mutated because a user might not be aware that it mutates the state of other variables as well. Is that correct?

One of the main challenges here is having a reasonable precise alias analysis. I would have to think through more examples. My worry is that there are many valid use cases for this. For example. your `change_first` example seems fine to me.

**Comment by boolean-light** [comment]
> I understand that the rule you're asking for is to warn if multiple variables point to the same data and that data gets mutated because a user might not be aware that it mutates the state of other variables as well. Is that correct?

Yes.

> your `change_first` example seems fine to me.

Realized this will be fine after revising my code, and probably this should be marked as one valid type.

I think it is fine for the narrow case where two different names are just pointing to the same "reference" and same binding, but I agree that we should consider more examples before implementing this.
